external/fv3fit/fv3fit/emulation/transforms.py

```diff
@@ -0,0 +1,83 @@
+import dataclasses
+from typing import List, Set
+
+import tensorflow as tf
+from typing_extensions import Protocol
+from fv3fit.emulation.types import TensorDict
+
+
+class TensorTransform(Protocol):
+    def forward(self, x: TensorDict) -> TensorDict:
+        pass
+
+    def backward(self, x: TensorDict) -> TensorDict:
+        pass
+
+    def backward_names(self, requested_names: Set[str]) -> Set[str]:
+        """input names needed to compute the requested transformed names
+
+        Needed for data loading purposes.
+        
+        """
+        pass
+
+
+@dataclasses.dataclass
+class LogTransform:
+    """A univariate transformation for::
+
+        y := log(max(x,epsilon))
+        x : = exp(x)
+
+    This is not strictly a bijection because of the quashing at epsilon.
+
+    Attributes:
+        epsilon: the size of the log transform
+    """
+
+    epsilon: float = 1e-30
+
+    def forward(self, x: tf.Tensor) -> tf.Tensor:
+        return tf.math.log(tf.maximum(x, self.epsilon))
+
+    def backward(self, x: tf.Tensor) -> tf.Tensor:
+        return tf.math.exp(x)
+
+
+@dataclasses.dataclass
+class TransformedVariableConfig:
+    """A user facing implementation"""
+
+    source: str
+    to: str
+    transform: LogTransform
```

Comment:
From a configuration perspective, this is a bit opaque.  If we add more transforms, we should switch to something similar to what we do for Optimizers, etc., with a dataclass taking `name` and `kwargs`.

```
source: field_in
to: transformed_field
transform:
  epsilon: 0.000...001
```
--> to
```
source: field_in
to: transformed_field
transform:
  name: LogTransform
  kwargs: {}
```